### PR TITLE
For review only

### PR DESCRIPTION
**Pull request by jtsaito**
This PR splits up the code in master into reviewable commits. The files in this branch are a subset of the master (this branch contains the source files only and omits `scripts`, `.travis.yml`, `vendor/`, `website/` and other boring files).

You can review and comment commit by commit. The most interesting commit for review might be the "Add repository test" commit.

I tried to stick close to the naming convention used by the [AWS provider](https://github.com/terraform-providers/terraform-provider-aws) which is supposed to be a model implementation by Hashicorp.

### CODE REVIEW

File: travisci/config.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+type Config struct {
+	GithubOwner  string // github organisation or user
+	GithubToken  string // private github token
+	TravisAPIURL string // travis api (pro or default)
+}
+
+// Client configures and returns a fully initialized Travis-CI client
+func (c *Config) Client() (*travis.Client, error) {
+
+	if travisClients == nil {
```
Comment (jansiwy) [inline travisci/config.go]: AFAICS this is not thread-safe... or how to call it here? gopher-safe

as you are modifying a module variable, you might need a lock around this

File: travisci/config.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
+package travisci
+
+import travis "github.com/jtsaito/go-travis"
+
+// Cache client instance per Travis-CI provider instance.
+// Significantly reduces risk of exceeding GitHub's abuse limit.
+var travisClients map[string]map[string]map[string]*travis.Client
```
Comment (jansiwy) [inline travisci/config.go]: what about `map[Config]*travis.Client` instead?

File: travisci/data_source_travisci_sync.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+				Optional: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func dataSourceSyncRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	client, err := config.Client()
```
Comment (jansiwy) [inline travisci/data_source_travisci_sync.go]: `err` not handled

File: travisci/data_source_travisci_sync.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+	var localSyncedAt, remoteSyncedAt time.Time
+	l := d.Get("synced_at").(string)
+	if l == "" {
+		localSyncedAt = time.Unix(0, 0)
+	} else {
+		localSyncedAt, err = time.Parse("2006-01-02T15:04:05Z", l)
```
Comment (jansiwy) [inline travisci/data_source_travisci_sync.go]: i've read the documentation, but I still don't believe that there is a magic reference date :-/

File: travisci/data_source_travisci_sync.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+		if err != nil {
+			return err
+		}
+	}
+
+	user, _, err := client.Users.GetAuthenticated()
```
Comment (jansiwy) [inline travisci/data_source_travisci_sync.go]: `err` redefined?

File: travisci/data_source_travisci_sync.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+	if localSyncedAt.Before(remoteSyncedAt) {
+		s, err := dataSourceSyncAndWaitForUpdate(client)
+		if err != nil {
+			return nil
+		}
+		d.Set("synced_at", s)
+	}
+
+	d.SetId("static")
```
Comment (jansiwy) [inline travisci/data_source_travisci_sync.go]: I've seen in other data sources, that they've used the date as ID... what do you think?

File: travisci/data_source_travisci_sync.go [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+func dataSourceSyncAndWaitForUpdate(client *travis.Client) (string, error) {
+	_, err := client.Users.Sync()
+	if err != nil {
+		if !strings.Contains(err.Error(), "409") {
+			return "", err
+		}
+	}
+
+	for waitSeconds := 10 * time.Second; waitSeconds <= 60*time.Second; waitSeconds *= 2 {
```
Comment (jansiwy) [inline travisci/data_source_travisci_sync.go]: why exponential back-off? it's not error handling. it's expected waiting.

File: travisci/data_source_travisci_sync.go [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+	return nil
+}
+
+func dataSourceSyncAndWaitForUpdate(client *travis.Client) (string, error) {
+	_, err := client.Users.Sync()
+	if err != nil {
+		if !strings.Contains(err.Error(), "409") {
+			return "", err
+		}
```
Comment (jansiwy) [inline travisci/data_source_travisci_sync.go]: what if there is any other error?

**Comment by jtsaito** [comment]
Closing this in favour of https://github.com/babbel/terraform-provider-travisci/pull/15

**Pull request closed without merging**